src/zenml/materializers/base_materializer.py

```diff
@@ -77,23 +88,22 @@ def handle_input(self, data_type: Type[Any]) -> Any:
             Any object that is to be passed into the relevant artifact in the
             step.
         """
-        # TODO [ENG-140]: Add type checking for materializer handle_input
-        # if data_type not in self.ASSOCIATED_TYPES:
-        #     raise ValueError(
-        #         f"Data type {data_type} not supported by materializer "
-        #         f"{self.__name__}. Supported types: {self.ASSOCIATED_TYPES}"
-        #     )
+        if not self._can_handle_type(data_type):
+            raise TypeError(
+                f"Unable to handle type {data_type}. {self.__class__.__name__} "
+                f"can only read artifacts to the following types: "
+                f"{self.ASSOCIATED_TYPES}."
+            )
 
     def handle_return(self, data: Any) -> None:
         """Write logic here to handle return of the step function.
 
         Args:
             data: Any object that is specified as an input artifact of the step.
         """
-        # TODO [ENG-141]: Put proper type checking
-        # if data_type not in self.ASSOCIATED_TYPES:
-        #     raise ValueError(
-        #         f"Data type {data_type} not supported by materializer "
-        #         f"{self.__class__.__name__}. Supported types: "
-        #         f"{self.ASSOCIATED_TYPES}"
-        #     )
+        data_type = type(data)
```

Comment:
Having the input and return checks here mean that we need to call these methods of the base class when we are inheriting from it and create a new materializer. I see that we already do that for our previously implemented materialized with `super().handle_input(data_type)`. However, do you think it might be a bit cleaner (or perhaps less risky in the case someone forgets to write it) for our users if we were to wrap these two methods with the checks?

Reply:
Mhm can you explain how you would wrap those methods? To me calling the super method seems like a reasonable thing (and materializer subclasses can for example don't call it on purpose if they want to raise custom error messages instead)

Reply:
On second thought, wrapping it in the class might be a bit redundant. I was just slightly concerned due to the fact that type checking only works if you remember to put the super method in the method itself. In order to counter that, we could do the checks before the actual call of the `handle_input` in the `resolve_input_artifact` of the `_FunctionExecutor` however this way you only do type checking within the executors and I am not sure which one I favor.

Reply:
I feel like it's on us to make sure we either
- call the super method
- do manual error checking
in all materializer subclasses. Worst case is it crashes with an error message that is a little less descriptive, but I feel like duplicating checks in many places is more of a burden to maintain